# Cats and Dogs Image Classification

In [1]:
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Convolution2D, MaxPooling2D
from keras.layers import Activation, Dropout, Flatten, Dense


# dimensions of our images.
img_width, img_height = 150, 150

train_data_dir = 'data/train'
validation_data_dir = 'data/validation'
nb_train_samples = 2000
nb_validation_samples = 802
nb_epoch = 50


model = Sequential()
model.add(Convolution2D(32, 3, 3, input_shape=(3, img_width, img_height)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Convolution2D(64, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Flatten())
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

# this is the augmentation configuration we will use for training
train_datagen = ImageDataGenerator(
        rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)

# this is the augmentation configuration we will use for testing:
# only rescaling
test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
        train_data_dir,
        target_size=(img_width, img_height),
        batch_size=32,
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_data_dir,
        target_size=(img_width, img_height),
        batch_size=32,
        class_mode='binary')

model.fit_generator(
        train_generator,
        samples_per_epoch=nb_train_samples,
        nb_epoch=nb_epoch,
        validation_data=validation_generator,
        nb_val_samples=nb_validation_samples)

model.save_weights('second_try.h5')


Using Theano backend.
Using gpu device 0: GeForce GTX 980M (CNMeM is enabled with initial size: 80.0% of memory, cuDNN not available)


Found 2000 images belonging to 2 classes.
Found 802 images belonging to 2 classes.
Epoch 1/50
2000/2000 [==============================] - 40s - loss: 0.7039 - acc: 0.5225 - val_loss: 0.6667 - val_acc: 0.5923
Epoch 2/50
2000/2000 [==============================] - 21s - loss: 0.6790 - acc: 0.5750 - val_loss: 0.6370 - val_acc: 0.6372
Epoch 3/50
2000/2000 [==============================] - 21s - loss: 0.6437 - acc: 0.6275 - val_loss: 0.6265 - val_acc: 0.6571
Epoch 4/50
2000/2000 [==============================] - 22s - loss: 0.6036 - acc: 0.6810 - val_loss: 0.6137 - val_acc: 0.6471
Epoch 5/50
2000/2000 [==============================] - 21s - loss: 0.5939 - acc: 0.6820 - val_loss: 0.5640 - val_acc: 0.7045
Epoch 6/50
2000/2000 [==============================] - 21s - loss: 0.5790 - acc: 0.6920 - val_loss: 0.6139 - val_acc: 0.6796
Epoch 7/50
2000/2000 [==============================] - 22s - loss: 0.5505 - acc: 0.7245 - val_loss: 0.5564 - val_acc: 0.7195
Epoch 8/50
2000/2000 [=============

Found 2000 images belonging to 2 classes.
Found 802 images belonging to 2 classes.
Epoch 1/50
2000/2000 [==============================] - 625s - loss: 0.6958 - acc: 0.4935 - val_loss: 0.6933 - val_acc: 0.5000
Epoch 2/50
2000/2000 [==============================] - 576s - loss: 0.6931 - acc: 0.5250 - val_loss: 0.6780 - val_acc: 0.5723
Epoch 3/50
2000/2000 [==============================] - 577s - loss: 0.6906 - acc: 0.5675 - val_loss: 0.6786 - val_acc: 0.5586
Epoch 4/50
2000/2000 [==============================] - 577s - loss: 0.6772 - acc: 0.5695 - val_loss: 0.6517 - val_acc: 0.5823
Epoch 5/50
2000/2000 [==============================] - 577s - loss: 0.6595 - acc: 0.5960 - val_loss: 0.6458 - val_acc: 0.6384
Epoch 6/50
2000/2000 [==============================] - 577s - loss: 0.6377 - acc: 0.6440 - val_loss: 0.6168 - val_acc: 0.6584
Epoch 7/50
2000/2000 [==============================] - 577s - loss: 0.6240 - acc: 0.6530 - val_loss: 0.6368 - val_acc: 0.6347
Epoch 8/50
2000/2000 [==============================] - 578s - loss: 0.6139 - acc: 0.6550 - val_loss: 0.6215 - val_acc: 0.6596
Epoch 9/50
2000/2000 [==============================] - 582s - loss: 0.6101 - acc: 0.6845 - val_loss: 0.5955 - val_acc: 0.6758
Epoch 10/50
2000/2000 [==============================] - 582s - loss: 0.5880 - acc: 0.6895 - val_loss: 0.6037 - val_acc: 0.6683
Epoch 11/50
2000/2000 [==============================] - 582s - loss: 0.5791 - acc: 0.7015 - val_loss: 0.6910 - val_acc: 0.6584
Epoch 12/50
2000/2000 [==============================] - 583s - loss: 0.5791 - acc: 0.7025 - val_loss: 0.6373 - val_acc: 0.6870
Epoch 13/50
2000/2000 [==============================] - 583s - loss: 0.5790 - acc: 0.6990 - val_loss: 0.5747 - val_acc: 0.6845
Epoch 14/50
2000/2000 [==============================] - 586s - loss: 0.5609 - acc: 0.7140 - val_loss: 0.6807 - val_acc: 0.6820
Epoch 15/50
2000/2000 [==============================] - 583s - loss: 0.5654 - acc: 0.7020 - val_loss: 0.5897 - val_acc: 0.6596
Epoch 16/50
2000/2000 [==============================] - 651s - loss: 0.5613 - acc: 0.7205 - val_loss: 0.6887 - val_acc: 0.6858
Epoch 17/50
2000/2000 [==============================] - 669s - loss: 0.5564 - acc: 0.7130 - val_loss: 0.7115 - val_acc: 0.6509
Epoch 18/50
2000/2000 [==============================] - 663s - loss: 0.5521 - acc: 0.7270 - val_loss: 0.6200 - val_acc: 0.6696
Epoch 19/50
2000/2000 [==============================] - 660s - loss: 0.5322 - acc: 0.7240 - val_loss: 0.5523 - val_acc: 0.6958
Epoch 20/50
2000/2000 [==============================] - 656s - loss: 0.5534 - acc: 0.7185 - val_loss: 0.5712 - val_acc: 0.7045
Epoch 21/50
2000/2000 [==============================] - 665s - loss: 0.5401 - acc: 0.7310 - val_loss: 0.5755 - val_acc: 0.6995
Epoch 22/50
2000/2000 [==============================] - 666s - loss: 0.5703 - acc: 0.7015 - val_loss: 0.5664 - val_acc: 0.7207
Epoch 23/50
2000/2000 [==============================] - 665s - loss: 0.5466 - acc: 0.7330 - val_loss: 0.5992 - val_acc: 0.6895
Epoch 24/50
2000/2000 [==============================] - 601s - loss: 0.5355 - acc: 0.7405 - val_loss: 0.5486 - val_acc: 0.7032
Epoch 25/50
2000/2000 [==============================] - 584s - loss: 0.5243 - acc: 0.7375 - val_loss: 0.6036 - val_acc: 0.6746
Epoch 26/50
2000/2000 [==============================] - 583s - loss: 0.5350 - acc: 0.7400 - val_loss: 0.5628 - val_acc: 0.7057
Epoch 27/50
2000/2000 [==============================] - 584s - loss: 0.5071 - acc: 0.7545 - val_loss: 0.6793 - val_acc: 0.6808
Epoch 28/50
2000/2000 [==============================] - 583s - loss: 0.5127 - acc: 0.7425 - val_loss: 0.5228 - val_acc: 0.7369
Epoch 29/50
2000/2000 [==============================] - 582s - loss: 0.4955 - acc: 0.7610 - val_loss: 0.5606 - val_acc: 0.7456
Epoch 30/50
2000/2000 [==============================] - 581s - loss: 0.5317 - acc: 0.7545 - val_loss: 0.5534 - val_acc: 0.7132
Epoch 31/50
2000/2000 [==============================] - 583s - loss: 0.5066 - acc: 0.7650 - val_loss: 0.5145 - val_acc: 0.7506
Epoch 32/50
2000/2000 [==============================] - 582s - loss: 0.4943 - acc: 0.7480 - val_loss: 0.5258 - val_acc: 0.7232
Epoch 33/50
2000/2000 [==============================] - 583s - loss: 0.4892 - acc: 0.7585 - val_loss: 0.6265 - val_acc: 0.6459
Epoch 34/50
2000/2000 [==============================] - 582s - loss: 0.5010 - acc: 0.7540 - val_loss: 0.5284 - val_acc: 0.7444
Epoch 35/50
2000/2000 [==============================] - 583s - loss: 0.4813 - acc: 0.7770 - val_loss: 0.6567 - val_acc: 0.7219
Epoch 36/50
2000/2000 [==============================] - 582s - loss: 0.4818 - acc: 0.7690 - val_loss: 0.5966 - val_acc: 0.7232
Epoch 37/50
2000/2000 [==============================] - 583s - loss: 0.4744 - acc: 0.7820 - val_loss: 0.5232 - val_acc: 0.7219
Epoch 38/50
2000/2000 [==============================] - 582s - loss: 0.4619 - acc: 0.7890 - val_loss: 0.5639 - val_acc: 0.7107
Epoch 39/50
2000/2000 [==============================] - 583s - loss: 0.4793 - acc: 0.7700 - val_loss: 0.5472 - val_acc: 0.7132
Epoch 40/50
2000/2000 [==============================] - 583s - loss: 0.5117 - acc: 0.7570 - val_loss: 0.5519 - val_acc: 0.7232
Epoch 41/50
2000/2000 [==============================] - 583s - loss: 0.4882 - acc: 0.7825 - val_loss: 0.6979 - val_acc: 0.6958
Epoch 42/50
2000/2000 [==============================] - 584s - loss: 0.4855 - acc: 0.7585 - val_loss: 0.5974 - val_acc: 0.7481
Epoch 43/50
2000/2000 [==============================] - 584s - loss: 0.5185 - acc: 0.7520 - val_loss: 0.5475 - val_acc: 0.7357
Epoch 44/50
2000/2000 [==============================] - 584s - loss: 0.4828 - acc: 0.7800 - val_loss: 0.5648 - val_acc: 0.7007
Epoch 45/50
2000/2000 [==============================] - 583s - loss: 0.4893 - acc: 0.7700 - val_loss: 0.5661 - val_acc: 0.7319
Epoch 46/50
2000/2000 [==============================] - 583s - loss: 0.4531 - acc: 0.7955 - val_loss: 0.5795 - val_acc: 0.7531
Epoch 47/50
2000/2000 [==============================] - 583s - loss: 0.4585 - acc: 0.7900 - val_loss: 0.5641 - val_acc: 0.7406
Epoch 48/50
2000/2000 [==============================] - 584s - loss: 0.4630 - acc: 0.7960 - val_loss: 0.6566 - val_acc: 0.7294
Epoch 49/50
2000/2000 [==============================] - 584s - loss: 0.4686 - acc: 0.7840 - val_loss: 0.5976 - val_acc: 0.7481
Epoch 50/50
2000/2000 [==============================] - 584s - loss: 0.4471 - acc: 0.7990 - val_loss: 0.5456 - val_acc: 0.7556